controllers/topology/reconcile_state.go

```diff
@@ -62,17 +66,158 @@ func (r *ClusterReconciler) reconcileCluster(ctx context.Context, current, desir
+	}
+
+	log.Info("creating", md.object.GroupVersionKind().String(), md.object.GetName())
+	if err := r.Client.Create(ctx, md.object.DeepCopy()); err != nil {
+		return errors.Wrapf(err, "failed to create the %s %s", md.object.GroupVersionKind(), md.object.Name)
+	}
+	return nil
+}
+
+// updateMachineDeployment updates a MachineDeployment. Also rotates the corresponding Templates if necessary.
+func (r *ClusterReconciler) updateMachineDeployment(ctx context.Context, clusterName string, currentMD, desiredMD machineDeploymentTopologyState) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	cleanupOldInfrastructureTemplate, err := r.reconcileReferencedTemplate(ctx, reconcileReferencedTemplateInput{
+		ref:     &desiredMD.object.Spec.Template.Spec.InfrastructureRef,
+		current: currentMD.infrastructureMachineTemplate,
+		desired: desiredMD.infrastructureMachineTemplate,
+		templateNamer: func() string {
+			// TODO: FIXME: Move the name generation into a shared func after #5070 has been merged
+			return names.SimpleNameGenerator.GenerateName(fmt.Sprintf("%s-%s-infra-", clusterName, desiredMD.object.Name))
+		},
+		compatibilityChecker: checkReferencedObjectsAreCompatible,
+	})
+	if err != nil {
+		return errors.Wrapf(err, "failed to update the %s %s", currentMD.object.GroupVersionKind(), currentMD.object.Name)
+	}
+
+	cleanupOldBootstrapTemplate, err := r.reconcileReferencedTemplate(ctx, reconcileReferencedTemplateInput{
+		ref:     desiredMD.object.Spec.Template.Spec.Bootstrap.ConfigRef,
+		current: currentMD.bootstrapTemplate,
+		desired: desiredMD.bootstrapTemplate,
+		templateNamer: func() string {
+			// TODO: FIXME: Move the name generation into a shared func after #5070 has been merged
+			return names.SimpleNameGenerator.GenerateName(fmt.Sprintf("%s-%s-bootstrap-", clusterName, desiredMD.object.Name))
+		},
+		compatibilityChecker: checkReferencedObjectsAreInTheSameNamespace,
+	})
+	if err != nil {
+		return errors.Wrapf(err, "failed to update the %s %s", currentMD.object.GroupVersionKind(), currentMD.object.Name)
+	}
+
+	// Check differences between current and desired MachineDeployment, and eventually patch the current object.
+	// TODO @fabrizio Do we also need the DeepCopy here?
+	patchHelper, err := newMergePatchHelper(currentMD.object, desiredMD.object, r.Client)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create patch helper for %s %s", currentMD.object.GroupVersionKind(), currentMD.object.Name)
+	}
+	if patchHelper.HasChanges() {
+		log.Info("updating", currentMD.object.GroupVersionKind().String(), currentMD.object.GetName())
+		// TODO @fabrizio I see a few {"metadata":{"creationTimestamp":null}} patches in unit tests, I think it will also be there
+		// in "reality", any follow-up tasks?
+		if err := patchHelper.Patch(ctx); err != nil {
+			return errors.Wrapf(err, "failed to update the %s %s", currentMD.object.GroupVersionKind(), currentMD.object.Kind)
```

Comment:
nit
```suggestion
			return errors.Wrapf(err, "failed to update %s %s", currentMD.object.GroupVersionKind(), currentMD.object.Kind)
```
(same for other error messages)